### PR TITLE
guess image formats

### PR DESCRIPTION
**Pull request by robtfm**
# Objective

provide an option to use `image::guess_format` to determine the image format, and make it the default. 

useful when loading images of unknown format (eg from urls without extension/mimetype), and more robust when loading mislabelled files.

### CODE REVIEW

File: crates/bevy_render/src/texture/image.rs
Comment (alice-i-cecile) [inline crates/bevy_render/src/texture/image.rs:103]:
```suggestion
    /// Attempts to convert from [`image::ImageFormat`].
    ///
    /// Unsupported formats will become [`None`].
    pub fn from_image_crate_format(format: image::ImageFormat) -> Option<Self> {
```

Comment (alice-i-cecile) [reply crates/bevy_render/src/texture/image.rs:103]: I'd also be in favor updating the docs on the sibling methods to match in this PR.

File: crates/bevy_render/src/texture/image_loader.rs
Comment (alice-i-cecile) [inline crates/bevy_render/src/texture/image_loader.rs:52]: Please add doc strings to this enum while we're here.

**Comment by robtfm** [comment]
> 4\. On some level I'd like to check and warn if the extension does not match the inferred type. However I'm not sure if this is the right spot for it: it will slow down loading in the happy path.

we could do that here, i don't think the perf impact would be noticeable (when compared to loading the binary data, converting internally, pushing it to gpu, etc) . 

if so i'd like to add another extension (say `.image`) which would never trigger the warning. i'm downloading files with no extension where all i know is that they are images, so i have to hack some recognised extension onto the "path" (which is acually a urn plus base url encoded into a path) to get the loader to pick up the asset, and the warning would be noisy and useless for me in that case.

File: crates/bevy_render/src/texture/image_loader.rs
Comment (tguichaoua) [inline crates/bevy_render/src/texture/image_loader.rs:112]:
```suggestion
                            path: load_context.path().display().to_string(),
```

`to_string` can be used here.

File: crates/bevy_render/src/texture/image_loader.rs
Comment (tguichaoua) [inline crates/bevy_render/src/texture/image_loader.rs:118]:
```suggestion
                            path: load_context.path().display().to_string(),
```
`to_string` can be used here.

**Review by tguichaoua — CHANGES_REQUESTED** [review]

File: crates/bevy_render/src/texture/image_loader.rs
Comment (robtfm) [reply crates/bevy_render/src/texture/image_loader.rs]: thanks, i also modified the one i copypasted from

**Comment by alice-i-cecile** [comment]
@robtfm that makes sense! Let's add that check and this special extension type here.

**Comment by rparrett** [comment]
> so i have to hack some recognised extension onto the "path"

That is still necessary after #10153?

**Comment by robtfm** [comment]
> That is still necessary after #10153?

Nice, I hadn’t seen that - I’ll check there’s no issues with using extensionless and then remove the “.image” extension

File: crates/bevy_render/src/texture/image.rs
Comment (bushrat011899) [inline crates/bevy_render/src/texture/image.rs:115]: Minor nit, could this be replaced with an implementation of `TryFrom` instead?

<details>
  <summary>Example</summary>
  
```rust
/// Error returned when converting between Bevy and [image] image format types.
#[derive(Debug, Error)]
#[error("Image format '{image:?}' maps to '{bevy:?}'")]
pub struct UnsupportedImageFormatMapping {
    /// Image format type from [image]
    pub image: Option<image::ImageFormat>,
    /// Bevy's image format type
    pub bevy: Option<ImageFormat>,
}

impl TryFrom<image::ImageFormat> for ImageFormat {
    type Error = UnsupportedImageFormatMapping;

    fn try_from(format: image::ImageFormat) -> Result<Self, Self::Error> {
        let format = match format {
            image::ImageFormat::Avif => ImageFormat::Avif,
            image::ImageFormat::Bmp => ImageFormat::Bmp,
            // ...
            _ => {
                return Err(Self::Error {
                    image: Some(format),
                    bevy: None,
                })
            }
        };

        Ok(format)
    }
}

impl TryFrom<ImageFormat> for image::ImageFormat {
    type Error = UnsupportedImageFormatMapping;

    fn try_from(format: ImageFormat) -> Result<Self, Self::Error> {
        let format = match format {
            ImageFormat::Avif => image::ImageFormat::Avif,
            ImageFormat::Bmp => image::ImageFormat::Bmp,
            // ...
            _ => {
                return Err(Self::Error {
                    image: None,
                    bevy: Some(format),
                })
            }
        };

        Ok(format)
    }
}

impl ImageFormat {
    // ...

    /// Attempts to convert to [`image::ImageFormat`].
    ///
    /// Unsupported formats will become [`None`].
    pub fn as_image_crate_format(&self) -> Option<image::ImageFormat> {
        (*self).try_into().ok()
    }

    /// Attempts to convert from [`image::ImageFormat`].
    ///
    /// Unsupported formats will become [`None`].
    pub fn from_image_crate_format(format: image::ImageFormat) -> Option<Self> {
        format.try_into().ok()
    }

    // ...
}
```
  
</details>

I personally think it's nicer to use the core traits when appropriate.

**Review by bushrat011899 — COMMENTED** [review]
Nice change! There's a little more work required to get something like:

```rust
commands.spawn(SpriteBundle {
    texture: asset_server.load("branding/bevy_bird_dark"),
    ..default()
});
```

Since there are multiple asset loaders registered which can handle an extensionless path for an `Image`. But that's beyond the scope for this PR.

**Comment by robtfm** [comment]
> Since there are multiple asset loaders registered which can handle an extensionless path for an `Image`. But that's beyond the scope for this PR.

right ... the built-in ones could probably be unified, but users might want to register other image loaders. i guess we may want a "default" loader for a given asset type (or possibly to try all the loaders?), but for now i'll leave the magic extension.

**Review by alice-i-cecile — APPROVED** [review]

**Review by janhohenheim — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@robtfm once merge conflicts are resolved I'll merge this in for you.

**Comment by alice-i-cecile** [comment]
@robtfm I'll merge this ASAP once merge conflicts are fixed.

**Comment by alice-i-cecile** [comment]
Completed in #13575

**Pull request closed without merging**
